src/Generator/Generators/CSharp/CSharpSources.cs

```diff
@@ -1580,98 +1580,127 @@ public void GenerateVTable(Class @class)
 
-            NewLine();
+                        AllocateNewVTables(@class, wrappedEntries, destructorOnly: false, "ManagedVTables");
 
-            if (hasVirtualDtor)
-                AllocateNewVTables(@class, wrappedEntries, destructorOnly: true);
+                        WriteLine($"initialized = true;");
+                        WriteLine($"VTables = new CppSharp.Runtime.VTables(originalVTables, methodDelegates);");
 
-            AllocateNewVTables(@class, wrappedEntries, destructorOnly: false);
+                        if (!hasVirtualDtor)
+                        { 
+                            WriteLine($"if ({destructorOnly})");
+                            WriteLineIndent("return VTables;");
+                        }
 
-            Write("__originalVTables = ");
-            SaveOriginalVTablePointers(@class);
-            UnindentAndWriteCloseBrace();
+                        UnindentAndWriteCloseBrace();
+                    }
+                    NewLine();
 
-            NewLine();
-            if (hasVirtualDtor)
-            {
-                WriteLine($"if ({destructorOnly})");
-                WriteOpenBraceAndIndent();
-                AssignNewVTableEntries(@class, "__ManagedVTablesDtorOnly");
-                UnindentAndWriteCloseBrace();
-                WriteLine("else");
-                WriteOpenBraceAndIndent();
-                AssignNewVTableEntries(@class, "__ManagedVTables");
-                UnindentAndWriteCloseBrace();
-            }
-            else
-            {
-                AssignNewVTableEntries(@class, "__ManagedVTables");
-            }
+                    if (hasVirtualDtor)
+                    {
+                        WriteLine($"if ({destructorOnly})");
+                        {
+                            WriteOpenBraceAndIndent();
+                            AssignNewVTableEntries(@class, "ManagedVTablesDtorOnly");
+                            UnindentAndWriteCloseBrace();
+                        }
+                        WriteLine("else");
+                        {
+                            WriteOpenBraceAndIndent();
+                            AssignNewVTableEntries(@class, "ManagedVTables");
+                            UnindentAndWriteCloseBrace();
+                        }
+                    }
+                    else
+                    {
+                        AssignNewVTableEntries(@class, "ManagedVTables");
+                    }
 
+                    WriteLine("return VTables;");
+                    UnindentAndWriteCloseBrace();
+                }
+            }
             UnindentAndWriteCloseBrace();
             NewLine();
+
+            if (!hasDynamicBase)
+                WriteLine("protected CppSharp.Runtime.VTables __vtables;");
+
+            WriteLines($@"
+            internal {(hasDynamicBase ? "override" : "virtual")} CppSharp.Runtime.VTables __VTables
+            {{ 
+                get {{
+                    if (__vtables.IsEmpty)
```

Comment:
Is this necessary? We initialize tables in constructors.

Reply:
`CreateInstance` and `GetOrCreateInstance` both have a `skipVTables` parameter which is used in some cases.